강좌 덮어쓰기 기능 추가

JuTaK97 opened the pull request:
- SearchFragment에서 에러 종류 판단해서 강의 중복 에러를 받으면 dialog를 띄워서 확인 누르면 is_force로 재수행


TODO 
- 지금은 SearchFragment에 moshi inject해서 dialog 띄우는 걸로 일단 해 놓았습니다.
- 문구 및 디자인 확정 필요합니다.
- 지금은 SearchFragment에만 적용해 놨는데, 강의 수정 및 커스텀 강의에도 적용할 지 고려해야 합니다. Custom 추가할 때는 다른 API를 쓰는 데, 서버와 얘기해봐야 합니다.

sanggggg committed:
temp for git

JuTaK97 committed:
NotificationFragment JetpackCompose 피드백 반영

sanggggg committed:
Merge branch 'develop' into jutak97/compose-notification-fragment

JuTaK97 committed:
중복 강의 선택지 추가

JuTaK97 committed:
Merge branch 'jutak97/compose-notification-fragment' of https://github.com/wafflestudio/SNUTT-android into jutak97/compose-notification-fragment

JuTaK97 committed:
lint 해결

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/data/MyLectureRepository.kt:
```diff
@@ -28,10 +28,11 @@ class MyLectureRepository @Inject constructor(
     val lastViewedTable = storage.lastViewedTable
 
-    fun addLecture(lectureId: String): Single<PostCustomLectureResults> {
+    fun addLecture(lectureId: String, is_forced: Boolean): Single<PostCustomLectureResults> {
```
[nit]
함수 인자나 변수 등등은 camelCase 를 컨벤션으로 가져가주세요!

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/lib/network/ApiOnError.kt:
```diff
@@ -227,11 +227,15 @@ class ApiOnError @Inject constructor(
-                        context.getString(R.string.error_lecture_time_overlap),
-                        Toast.LENGTH_SHORT
-                    ).show()
+//                    ErrorCode.LECTURE_TIME_OVERLAP -> {
+//                        restError.ext?.message?.let {
+//                            Toast.makeText(
+//                                context,
+//                                it,
+//                                Toast.LENGTH_SHORT
+//                            ).show()
+//                        }
+//                    }
```
주석 처리하지 않고 그대로 두거나 / 아예 삭제하는게 좋을 것 같아요.
현재 상황에 따라 달라지겠지만, 다른 오버랩 대응을 하지 않은 Lecture Add api 에서는 이 방식으로 핸들링 될 것 같아서?

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/lib/network/dto/PostLectureParams.kt:
```diff
@@ -1,7 +1,14 @@
 package com.wafflestudio.snutt2.lib.network.dto
 
+import com.squareup.moshi.Json
 import com.squareup.moshi.JsonClass
 
 // Refactor FIXME empty body
 @JsonClass(generateAdapter = true)
 class PostLectureParams()
+
+// 임시로 여기 추가
+@JsonClass(generateAdapter = true)
+data class PostLectureForce(
+    @Json(name = "is_forced") var id: Boolean
+)
```
이 is_forced 필드는 그냥 PostLectureParams 에 추가하고, PostLectureForce 대신에 PostLectureParams 를 사용하면 될 것 같아요

맥락이 PostLecture api 를 보낼 때는 이 body 를 보내겠다 -> PostLectureParams 라서?
+id 필드명 isForced 로 바꿔주세요

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchFragment.kt:
```diff
@@ -81,10 +89,22 @@ class SearchFragment : BaseFragment() {
+                                    val restError: RestError? = error.response()?.errorBody()?.string()?.let {
+                                        moshi.adapter(RestError::class.java).fromJson(it)
+                                    }
+                                    if (restError?.code == 0x300C) {
```
너무 300c 가 적나라 하게 박혀있는 것 같은데 `ErrorCode.어쩌구` 로 바꿔서 맥락을 전달해주세요
Enum 변환이 어려운 거였다면 일단 그대로 두고 FIXME 주석만 달아주세요! (moshi 여기서 안 쓰고 interceptor 단에서 쓰게 바꾸면서 제가 처리할게요)

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/timetable/SelectedTimetableViewModel.kt:
```diff
@@ -31,13 +31,13 @@ class SelectedTimetableViewModel @Inject constructor(
         get() = settingsRepository.tableTrimParam
 
-    fun toggleLecture(lecture: LectureDto): Completable {
+    fun toggleLecture(lecture: LectureDto, is_force: Boolean): Completable {
```
지금 생각하니.. 제가 작성했지만 toggleLecture 가 아니라 add remove 로 명시적인 두 함수를 만드는게 좋았을 것 같다는 생각이 드네요

매 호출에 대해서 이게 실제로 어떤 영향을 미치는 지 (삭제인가 추가인가) 알 수가 없어서 코드 보면서 조금 헷갈렸던 것 같아요
+isForced 필드는 삭제에 대해서는 영향력이 없는 필드인데 같이 들어가서 불필요한 맥락이 더 담기는 것 같기도 하네요

FIXME 달아주시면 제가 나중에 고치도록 하겠습니다, 아니면 주현님이 고쳐주셔도 좋을 것 같아요

sanggggg replied:
```suggestion
    // FIXME: @sangggg toggle 대신 add/remove 명시적으로 호출하기
    fun toggleLecture(lecture: LectureDto, is_force: Boolean): Completable {
```

JuTaK97 committed:
함수 인자 camelCase로 변경

https://github.com/wafflestudio/SNUTT-android/pull/50#discussion_r846648528

JuTaK97 committed:
ErrorCode.LECTURE_TIME_OVERLAP 주석 제거

추후 다른 overlap 경우에서 사용할 수도 있음

JuTaK97 committed:
PostLectureParams dto 수정

필드명 is_forced는 서버에서 그렇게 받도록 되어 있습니다.

JuTaK97 committed:
Merge branch 'develop' into jutak97/overwrapping-lecture

# Conflicts:
#	app/src/main/java/com/wafflestudio/snutt2/lib/network/ApiOnError.kt
#	app/src/main/java/com/wafflestudio/snutt2/manager/NotificationsRepository.kt
#	app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsFragment.kt
#	app/src/main/java/com/wafflestudio/snutt2/views/logged_in/notifications/NotificationsViewModel.kt

JuTaK97 committed:
NotificationsAdapter 임시 추가

Notification쪽 Jetpack Compose 변경으로 인해 삭제된 부분이지만, 이 브랜치에서 notification쪽 변경을 없애기 위해 임시로 옛날 코드 추가

JuTaK97 committed:
ErrorParsingCallAdpater 사용

https://github.com/wafflestudio/SNUTT-android/pull/56#issue-1198912970

JuTaK97 committed:
lint 정리

sanggggg committed:
Update app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/timetable/SelectedTimetableViewModel.kt

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchFragment.kt:
```diff
@@ -221,4 +239,24 @@ class SearchFragment : BaseFragment() {
+        val alert = AlertDialog.Builder(requireContext())
+            .setTitle("강좌 덮어씌우기")
+            .setMessage(message)
+            .setPositiveButton("조와용") { _, _ ->
```
문구만 마무리 해서 반영하고 나서 머지하시죠

JuTaK97 committed:
강의 추가/수정 시 버그 수정

강의 추가 및 수정 시 한번 에러(시간 겹침 등)가 나면 완료 버튼이 비활성화 되는 문제 해결

JuTaK97 committed:
Merge branch 'jutak97/fix-bug-customLectureDetail' into jutak97/overwrapping-lecture

JuTaK97 committed:
커스텀 강의 중복 시 선택지 추가

JuTaK97 committed:
lint

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/home/search/SearchFragment.kt:
```diff
@@ -53,6 +57,9 @@ class SearchFragment : BaseFragment() {
     @Inject
     lateinit var snuttUrls: SnuttUrls
 
+    @Inject
+    lateinit var moshi: Moshi
+
```
필요 없는 inject 지우기

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/lecture_detail/CustomLectureDetailFragment.kt:
```diff
@@ -50,6 +51,8 @@ class CustomLectureDetailFragment : BaseFragment() {
     private var editable = false
     private var add = false
 
+    private val LECTURE_TIME_OVERLAP = 0x300C
```
https://github.com/wafflestudio/SNUTT-android/blob/fa32dfd3f80a5e5386ddf7f5522b4a4ead905ff2/app/src/main/java/com/wafflestudio/snutt2/lib/network/ApiOnError.kt#L323
여기 정의되있는 값을 사용해주세요

sanggggg commented on app/src/main/java/com/wafflestudio/snutt2/views/logged_in/lecture_detail/CustomLectureDetailFragment.kt:
```diff
@@ -115,28 +118,30 @@ class CustomLectureDetailFragment : BaseFragment() {
+                        when (error) {
+                            is ErrorParsedHttpException -> {
+                                if (error.errorDTO?.code == LECTURE_TIME_OVERLAP) {
+                                    overwriteCustomLectureDialog(error.errorDTO.ext!!["confirm_message"])
```
!! 대신 디폴트 문구를 `"-"` 으로 넣는 식으로 처리하는게 크래시 방지용으로 좋을 것 같아요

JuTaK97 committed:
피드백 반영
